### ISSUE TITLE
[Micro Frontend] Duplicate class name if there is multiple styled-components

### ISSUE THREAD

**Issue by soulwu** [styled-components/styled-components#2974]
I use [qiankun](https://github.com/umijs/qiankun) to make my micro-frontend app.
I use `styled-components` for css-in-js.

If I mount multiple sub apps at the same time, and these sub apps use styled-components respectivly, the generated class name would be conflict.

Is there a way to config different seed to generate class name?

**Comment by probablyup** [comment]
Do you have a live example of this? The odds of a styling collision are very low unless there are multiple instances of styled-components on the page at once and both are rendering the same components in separate subtrees.

**Comment by casperle** [comment]
@soulwu Maybe the this comment can help you https://github.com/styled-components/styled-components/issues/1789#issuecomment-577704969

We have similar issue, but with an SSR setup which looks like this:
![image](https://user-images.githubusercontent.com/8900497/72994597-ab659880-3df7-11ea-9779-7f1a4faedae4.png)

The client calls the orchestrator which calls two different endpoints where react components (fragments) are rendered. The orchestrator puts it together and sends it to the client where every fragment gets hydrated. Both fragments are wrapped into `ThemeProvider`.
On the client side we have only one instance of the `styled-components` because no warning is provided.
With this setup we still get the following error:
![image](https://user-images.githubusercontent.com/8900497/72995314-fcc25780-3df8-11ea-9e7f-4c0da9b28852.png)

Also babel plugin is used with SSR setup.

Do you have any idea what can cause this?

**Comment by soulwu** [comment]
> @soulwu Maybe the this comment can help you [#1789 (comment)](https://github.com/styled-components/styled-components/issues/1789#issuecomment-577704969)
> 
> We have similar issue, but with an SSR setup which looks like this:
> ![image](https://user-images.githubusercontent.com/8900497/72994597-ab659880-3df7-11ea-9779-7f1a4faedae4.png)
> 
> The client calls the orchestrator which calls two different endpoints where react components (fragments) are rendered. The orchestrator puts it together and sends it to the client where every fragment gets hydrated. Both fragments are wrapped into `ThemeProvider`.
> On the client side we have only one instance of the `styled-components` because no warning is provided.
> With this setup we still get the following error:
> ![image](https://user-images.githubusercontent.com/8900497/72995314-fcc25780-3df8-11ea-9e7f-4c0da9b28852.png)
> 
> Also babel plugin is used with SSR setup.
> 
> Do you have any idea what can cause this?

Great! I also found a babel plugin [babel-plugin-styled-components-css-namespace](https://github.com/QuickBase/babel-plugin-styled-components-css-namespace) can resolve this

**Comment by rajzik** [comment]
Is there any other workaround? Because Babel plugin will break parcel build. Or will be there solution from styled components?

**Comment by BartJanvanAssen** [comment]
> Is there any other workaround? Because Babel plugin will break parcel build. Or will be there solution from styled components?

Having the same issue, posted a issue on the plugin’s page regarding nodejs version. Any updates on different solutions?

**Comment by pdrvsky** [comment]
Yup, I can confirm mentioned issue also. For me it's possible to regenerate it when I have two microfrontends build with webpack. It seems like it generates same class names, but only in built apps. When I debug it locally, it works like a charm.

**Comment by casperle** [comment]
Finally we resolved those issues by importing modules directly from styled-components. In the original setup we re-exported all modules from styled-components and because of that the babel plugin was not working at all. The plugin is checking the imported modules location and if its not from styled-components its skipped.

**Comment by ExeAy** [comment]
We are using a micro front setup with single-spa mostly consisting of react micro front ends and we are also having this issue. It's not a common issue for us though. 

Using [babel-plugin-styled-components-css-namespace](https://github.com/QuickBase/babel-plugin-styled-components-css-namespace) did not work since they seem to have an [issue](https://github.com/QuickBase/babel-plugin-styled-components-css-namespace/issues/54) with newer node versions. However just implementing [babel-plugin-styled-components](https://github.com/styled-components/babel-plugin-styled-components) seems to have done the trick. Thanks @pdrvsky for the tip.

**Comment by rajzik** [comment]
Hey, guys, I have forked babel plugin and added namespace setting, currently testing if it works reliably but it seems so. 
You can take a look here: 
[babel-plugin-styled-components-namespace](https://github.com/rajzik/babel-plugin-styled-components-namespace)

**Comment by pgangwani** [comment]
@rajzik : Is there a chance we can merge with exisiting babel-plugin-styled-components ?

**Comment by rajzik** [comment]
@pgangwani : I can create PR for the original babel plugin.

**Comment by jfairley** [comment]
I just had a class name collision. My scenario is a custom React component library used by a CRA frontend.

- The component library uses `styled-components` and is built with `tsc` (TypeScript).
- The frontend is built with `create-react-app`.

Interestingly:

- The collision only happens after I build the static app (`react-scripts build`). No problem in dev (`react-scripts start`).
- Using the macro (`styled-components/macro`) fixed the problem, though I suspect / worry this only sidestepped the problem by changing the algorithm.

<img width="1225" alt="duplicate" src="https://user-images.githubusercontent.com/992483/80242770-21f67a00-8623-11ea-97de-b8634d7017f0.png">

**Comment by NunoCardoso** [comment]
@jfairley I had exactly the same problem, and switching to styled-components/macro solved it.

**Comment by jfairley** [comment]
@NunoCardoso in your webpack config, are you using the babel- or ts-loader. The macro refers to the babel macros plugin.

**Comment by Cecil0o0** [comment]
> I just had a class name collision. My scenario is a custom React component library used by a CRA frontend.
> 
> * The component library uses `styled-components` and is built with `tsc` (TypeScript).
> * The frontend is built with `create-react-app`.
> 
> Interestingly:
> 
> * The collision only happens after I build the static app (`react-scripts build`). No problem in dev (`react-scripts start`).
> * Using the macro (`styled-components/macro`) fixed the problem, though I suspect / worry this only sidestepped the problem by changing the algorithm.
> 
> <img alt="duplicate" width="1225" src="https://user-images.githubusercontent.com/992483/80242770-21f67a00-8623-11ea-97de-b8634d7017f0.png">

@jfairley I had exactly the same problem too.

I guess that class name collision caused by two `runtime` of styled-components in two version, as shown below.

![image](https://user-images.githubusercontent.com/19543313/87424529-ac40ef80-c60e-11ea-8257-0ef6eddbaa70.png)

So far, I use the same version of styled-components between two projects, and then packed them into one web app, which works in my case!

**Comment by KSABCarl** [comment]
We have the same problem in production but not in development. In our case the two projects have the same version of styled-components.

One is referencing a styled.div and the other a styled.svg:
<img width="519" alt="Skärmavbild 2020-09-09 kl  07 05 23" src="https://user-images.githubusercontent.com/43408202/92556755-fd660c00-f26a-11ea-9a94-6acefa28e75d.png">

<img width="586" alt="Skärmavbild 2020-09-09 kl  07 05 42" src="https://user-images.githubusercontent.com/43408202/92556757-fdfea280-f26a-11ea-99b9-720f26c1eca4.png">

Duplicate entries for the className "cVmQYF" appear in the inspector:
<img width="331" alt="Skärmavbild 2020-09-09 kl  07 05 54" src="https://user-images.githubusercontent.com/43408202/92556760-fe973900-f26a-11ea-99ed-36005e807fe9.png">

The styled-components are on the same version:
<img width="448" alt="Skärmavbild 2020-09-09 kl  07 07 55" src="https://user-images.githubusercontent.com/43408202/92556847-343c2200-f26b-11ea-9cef-f757301c6f76.png">


We're using ts-loader so the babel-plugin isn't an option. Are there any other work-arorunds? All our individual components are built at the same time but are mounted individually on the same page.

**Comment by burakilk** [comment]
Did you find anything about it ?

**Comment by ExeAy** [comment]
> We have the same problem in production but not in development. In our case the two components have the same version of styled-components.
> 
> One is referencing a styled.div and the other a styled.svg:
> <img alt="Skärmavbild 2020-09-09 kl 07 05 23" width="519" src="https://user-images.githubusercontent.com/43408202/92556755-fd660c00-f26a-11ea-9a94-6acefa28e75d.png">
> 
> <img alt="Skärmavbild 2020-09-09 kl 07 05 42" width="586" src="https://user-images.githubusercontent.com/43408202/92556757-fdfea280-f26a-11ea-99b9-720f26c1eca4.png">
> 
> Duplicate entries for the className "cVmQYF" appear in the inspector:
> <img alt="Skärmavbild 2020-09-09 kl 07 05 54" width="331" src="https://user-images.githubusercontent.com/43408202/92556760-fe973900-f26a-11ea-99ed-36005e807fe9.png">
> 
> The styled-components are on the same version:
> <img alt="Skärmavbild 2020-09-09 kl 07 07 55" width="448" src="https://user-images.githubusercontent.com/43408202/92556847-343c2200-f26b-11ea-9cef-f757301c6f76.png">
> 
> We're using ts-loader so the babel-plugin isn't an option. Are there any other work-arorunds? All our individual components are built at the same time but are mounted individually on the same page.

I don't know about work arounds for ts-loader but you can use typescript with babel (see [babels web page](https://babeljs.io/docs/en/babel-preset-typescript)). This issue resulted in that we converted a few of out micro frontends from ts-loader to babel.

**Comment by KSABCarl** [comment]
> Did you find anything about it ?

We solved it with [`typescript-plugin-styled-components`](https://github.com/Igorbek/typescript-plugin-styled-components) though were not sure if we like the verbosity in production code.

We have additional problems right now with ts-loader so perhaps we'll have to move to babel anyway.

**Comment by danseaman6** [comment]
> We are using a micro front setup with single-spa mostly consisting of react micro front ends and we are also having this issue. It's not a common issue for us though.
> 
> Using [babel-plugin-styled-components-css-namespace](https://github.com/QuickBase/babel-plugin-styled-components-css-namespace) did not work since they seem to have an [issue](https://github.com/QuickBase/babel-plugin-styled-components-css-namespace/issues/54) with newer node versions. However just implementing [babel-plugin-styled-components](https://github.com/styled-components/babel-plugin-styled-components) seems to have done the trick. Thanks @pdrvsky for the tip.


@ExeAy in response to that - I'm also using a single-spa app with react pages. How did you get that babel plugin added? `single-spa`'s version of react-app-rewired doesn't include `customize-cra`, and their implementation of it doesn't make it particularly easy to add `customize-cra` and use something like `useBabelRc()`

**Comment by mcroba** [comment]
> > Did you find anything about it ?
> 
> We solved it with [`typescript-plugin-styled-components`](https://github.com/Igorbek/typescript-plugin-styled-components) though were not sure if we like the verbosity in production code.
> 
> We have additional problems right now with ts-loader so perhaps we'll have to move to babel anyway.

@KSABCarl Were you able to modify the dynamic classname with this plugin? If yes, how have you done it? The getDisplayName just modify the static class (eg. sc-...)

**Comment by KSABCarl** [comment]
> > > Did you find anything about it ?
> > 
> > 
> > We solved it with [`typescript-plugin-styled-components`](https://github.com/Igorbek/typescript-plugin-styled-components) though were not sure if we like the verbosity in production code.
> > We have additional problems right now with ts-loader so perhaps we'll have to move to babel anyway.
> 
> @KSABCarl Were you able to modify the dynamic classname with this plugin? If yes, how have you done it? The getDisplayName just modify the static class (eg. sc-...)

@mcroba No, we didn't because we didn't want to have to maintain extra code just to make it less verbose.

**Comment by Cecil0o0** [comment]
https://styled-components.com/docs/tooling#namespace

This works for me!👏

**Comment by petermikitsh** [comment]
I came across this issue and can confirm adding the following babel config to each microfrontend resolved the issue:

```json
{
  "plugins": [
    [
      "babel-plugin-styled-components",
      {
        "displayName": true
      }
    ]
  ]
}
```

**Comment by andywampir** [comment]
See my [comment](https://github.com/styled-components/styled-components/issues/3457#issuecomment-857402287) in another issue, that might help you

**Comment by wdfinch** [comment]
Just went through this issue. After spending many hours on it here are my findings and solutions:

The root cause for me was having multiple instances of `styled-components` loaded. This can easily be confirmed by seeing if there are multiple style tags in the head that say `data-styled-version`.

If so you have a few options. The best option is to not load multiple instances of `styled-components`. Some sources of multiple loads are:

1. Having `styled-components` inlined by a module bundler (rollup in my case). This happened for me. I had multiple packages used by various micro-frontend apps and rollup was inline `styled-components` and therefore duplicating it. If so look for an `external` option or plugin to keep the module from being inlined.

2. Multiple loads between webpack apps. If you're using module federation, use the singleton option to avoid multiple instances being loaded.

If you cannot avoid loading multiple instances then you should use the `babel-plugin-styled-components` plugin with the namespace and/or display name settings. This will namespace the various generated class name hashes and therefore eliminate the problem. If you are having issues using the plugin in either `webpack` or `rollup`, make sure `babel` is the first tool to transpile the code. In my case, compiling the code with typescript first resulted in the plugin not applying the transformations. I found the `fork-ts-checker-webpack-plugin` and `rollup-plugin-fork-ts-checker` useful to still get type checking while using babel as my primary transpile tool.

**Comment by tu4mo** [comment]
Babel plugin does not help in this case, see https://github.com/styled-components/styled-components/issues/3440.

**Comment by wdfinch** [comment]
I see. Good point.

**Comment by kitten** [comment]
I'm not really caught up on this, and I don't really understand what is found to be so attractive for some in duplicating loads of dependencies, instead of having a monolithic build process, that ensures shared library code as far as possible... unless maybe Preact is used instead of React due to the reduced size.... Anyway!

You can use multiple versions of styled-components on the same page by altering `process.env.SC_ATTR` per build. I haven't tested this in a while, but it should still work just fine.

**Comment by wdfinch** [comment]
I think the issue is sometimes we can't ensure all our dependencies are on the same version. For example, let's say we have a micro frontend with many apps developed by many separate teams. Team A might be used styled-components v5 for their app while team B might be using v4 in their app. It might not be feasible for team B to upgrade immediately and development cannot stop until they upgrade. 

Another similar situation is when the next major version comes out someday. If there are 50 frontend apps all using styled-components, then it would be a big feat to upgrade all of them to the next major version immediately (assuming there are breaking changes). Instead, a gradual approach would be better.

For small teams working on single apps none of this matters. But for large teams with big frontend apps developed over long periods of time, these issue start to become very relevant. 50 apps are not some crazy number either. It's reasonable for large products to have this depending on how finely you break up your application.

**Comment by kitten** [comment]
@wdfinch Correct me if I'm wrong, but the `namespace` option should hopefully then resolve this limitation, right?

**Comment by wdfinch** [comment]
I don't think so. The issue is the hash class name with the actual styled does not have the namespace. The image below is from my micro frontend app using the latest version of styled-components. I'm using both the namespace and display name options. So the highlighted element on the left shows the component name as TableSearchContainer and the app is react-common. As you can see on the right, the class name generated with the styles does not contain the namespace or display name. So unless I'm missing something, it seems like these options are more for aesthetic reasons (easier debugging). 
 
![Screenshot from 2021-08-09 08-49-17](https://user-images.githubusercontent.com/36898266/128735438-d48cdeb6-6e5a-4118-a3ed-baf417335591.png)

https://github.com/styled-components/styled-components/issues/3440#issue-839574836

**Comment by kitten** [comment]
@wdfinch Yep, that's still intended. However, this is a hash based on component IDs and the contents of the CSS string, so collisions will be extremely unlikely given the namespace option, read: as unlikely as any other hash collision with 32 bits if no mistake is made in our hashing logic.

**Comment by wdfinch** [comment]
Gotcha. Ok then I guess the namespace is a suitable choice. Perhaps the docs could be updated to mention that the namespace option will be factored into the hash.

**Comment by cscheffauer** [comment]
> Just went through this issue. After spending many hours on it here are my findings and solutions:
> 
> The root cause for me was having multiple instances of `styled-components` loaded. This can easily be confirmed by seeing if there are multiple style tags in the head that say `data-styled-version`.
> 
> If so you have a few options. The best option is to not load multiple instances of `styled-components`. Some sources of multiple loads are:
> 
> 1. Having `styled-components` inlined by a module bundler (rollup in my case). This happened for me. I had multiple packages used by various micro-frontend apps and rollup was inlining `styled-components` and therefore duplicating it. If so, look for an `external` option or plugin to keep the module from being inlined.
> 2. Multiple loads between webpack apps. If you're using module federation, use the singleton option to avoid multiple instances being loaded.
> 
> If you cannot avoid loading multiple instances then you should use the `babel-plugin-styled-components` plugin with the namespace and/or display name settings. This will namespace the various generated class name hashes and therefore eliminate the problem. If you are having issues using the plugin in either `webpack` or `rollup`, make sure `babel` is the first tool to transpile the code. In my case, compiling the code with typescript first resulted in the plugin not applying the transformations. I found the `fork-ts-checker-webpack-plugin` and `rollup-plugin-fork-ts-checker` useful to still get type checking while using babel as my primary transpile tool.

Thanks so much - that saved our day ! 💯 
Just as a note - we added `externals: [nodeExternals()]` (from webpack-node-externals) to get it working.

**Comment by alexanderhanssen** [comment]
Just went through this issue myself, and for people who have made sure that they only have one instance of `styled-components` but still are experiencing issues, this comment helped me: https://github.com/styled-components/styled-components/issues/2170#issuecomment-434294742. My issue was having two entrypoints, both using styled-component and using the same vendor file, but styled-components was initialized twice at runtime.

https://styled-components.com/docs/faqs#running-multiple-applications-on-one-page this hasn't been updated with the splitChunks plugin way of doing optimization, and there was a pull request made to update the docs: https://github.com/styled-components/styled-components-website/pull/378 but it didn't merge. 

So long story short, i added the runtimeChunck to my optimization like this:
```js
optimization: {
		splitChunks: {
			cacheGroups: {
				defaultVendors: {
					test: /[\\/]node_modules[\\/]/,
					chunks: "all",
					name: "libs",
					enforce: true,
					minChunks: 2,
				},
			},
		},
		runtimeChunk: { name: "libs" },
	},
```

**Comment by totszwai** [comment]
> #2170 (comment)

Do you have a site or resources that could show how to bundle/exclude `styled-components` out of React single-spa apps? Could you share your `craco.config.js`?

**Comment by Statfine** [comment]
change styled-components to styled-components/macro work for me

**Comment by fcano-uz** [comment]
> @wdfinch Yep, that's still intended. However, this is a hash based on component IDs and the contents of the CSS string, so collisions will be extremely unlikely given the namespace option, read: as unlikely as any other hash collision with 32 bits if no mistake is made in our hashing logic.

We only have two microfrontends and this has happened the moment we updated styled-components in one of them. To fix it, **we downgraded styled-components to version 4** in one microfrontend and the bug went away.

When this bug happened, it happened in **two** of our components: Inputs and Selects. So maybe it's not extremely unlikely? Or we were very unlucky 😭 

This bug also only happens when building for production, so catching during development is almost impossible. I understand that this might be extremely unlikely, but we need a better guarantee than that. Having no guarantee that any class can break at any minute for any change in another app is.. well.. not ideal...

I though `namespace` was meant to do that, but looks like something merely aesthetical... So, correct me if I'm wrong, but I guess that means the documentation is misleading? Since it states that `namespace` can avoid this issue, but it can't? (I'm refering to this documentation https://styled-components.com/docs/tooling#namespace and I'm assuming it's not working based on this comment: https://github.com/styled-components/styled-components/issues/2974#issuecomment-895339520)

**Comment by michaldromi** [comment]
If anyone is still struggling with this issue:

We encountered the same problem, facing duplicate class names from different apps in Module Federation due to multiple versions of styled-components. Here's a breakdown:

* An internal package was transitioned to an ESM build using rollup.
* This led to [this error](https://github.com/styled-components/styled-components/issues/3700).
* We attempted to resolve it by configuring the output-interop to "compact". However, this wasn't possible because NX utilizes rollup 2.
* To address the error, we had to remove styled-components from MF's shared dependencies, which led to the issue of duplicate classes.

Ultimately, the resolution was using different versions of styled-components across the MF apps. No other approaches proved effective (including Webpack aliasing and babel-plugin-styled-components). This approach was inspired by a comment found here: [link](https://github.com/styled-components/styled-components/issues/2543#issuecomment-491659106).

**Comment by probablyup** [comment]
There is a vote in progress to potentially remove the babel macro due to lack of usage/interest, please vote if you care thanks. https://github.com/styled-components/styled-components/issues/4064#issuecomment-1737517494

**Comment by quantizor** [comment]
Closing — this issue is stale (2+ years without resolution) and relates to an older version or has been superseded by v6.x changes. If this is still reproducible on the latest v6.x release, please open a new issue with a minimal reproduction.
